### ISSUE TITLE
Sharing files between Finder and VirtualC64

### ISSUE THREAD

**Issue by bluecursor** [dirkwhoffmann/virtualc64#559]
Make a easy way to share files between a virtual machine and the OSX environment.

**Comment by Alessandro1970** [comment]
...for loading .prg too

**Comment by dirkwhoffmann** [comment]
Won't be able to do this considering my limited resources (aka spare time). Sorry for that.

**Comment by dirkwhoffmann** [comment]
While browsing the web for a „Compiling VICE for dummies“ guide this morning, I stumbled across the following forum post by Retrofan (in German, sorry for that):

https://www.forum64.de/index.php?thread/101181-vice-am-mac-versionen-tastaturlayout/&pageNo=1

> „‚Schade, dass Power64 nicht mehr weiterentwickelt wird - das war der beste C64 Emulator für den Mac.‘
> 
> Dem trauere ich nach wie vor auch ein wenig nach. Klasse war vor allem das Laufwerkshandling, was Zusatzprogramme überflüssig machte. Einfach eine leere Diskette erzeugen, PRGs aus dem Finder darauf ziehen (oder heraus) und speichern. Oder zwischen Laufwerk 8 und 9 einfach per Drag & Drop Files austauschen. Das war schon sehr komfortabel und kann bisher kein anderer Emulator. Für solche Features könnte ich gerne auf die 100%ige Emulation von exotischer C64-Zusatz-Hardware verzichten.“

Technically, it’s pretty difficult to add files to a GCR encoded disk directly which is the reason why VirtualC64 doesn’t offer such functionality (yet). Retrofan's remark made me think about it again, and I believe it is not that difficult to implement such a feature:

To add a file to an existing disk, we can do the following: 

1. Convert the current disk to a D64Archive
2. Convert the D64Archive to a T64Archive
3. Add the new file to the T64Archive
4. Recreate the disk from the (modified) T64Archive

The only missing piece is step 3, but the T64 format is simple enough to be manipulated (in contrast to the D64 format or GCR encoded disks). Steps 1, 2, and 4 are already there. 

The main point here, I think, comes down to user experience. How do we integrate such a feature into the existing GUI. I don’t want to reimplement Finder. It should be a lean solution.

**Comment by dirkwhoffmann** [comment]
I'd like to move forward in smaller steps here (as said above, I don't want to implement a second Finder). In my current prototype, directories are accepted for drag and drop. E.g., if the following directory is dragged in,

<img width="457" alt="Bildschirmfoto 2020-09-14 um 12 38 57" src="https://user-images.githubusercontent.com/12561945/93076310-6f818b00-f687-11ea-95d4-53797d5a5c29.png">

the following dialog comes up:

<img width="854" alt="Bildschirmfoto 2020-09-14 um 12 35 52" src="https://user-images.githubusercontent.com/12561945/93076337-78725c80-f687-11ea-905b-f2c448c19ded.png">

After inserting the disk in drive 8, it can be used as every standard C64 disk can be: 

<img width="970" alt="Bildschirmfoto 2020-09-14 um 12 39 30" src="https://user-images.githubusercontent.com/12561945/93076532-c5563300-f687-11ea-9e65-2536b125512c.png">

Note: Neither is the disk linked to the original directory in any way (this is impossible with the way VirtualC64 is designed), nor do we have control over the order of files.

But it's a neat and easy way to get arbitrary files into the C64.

**Comment by bluecursor** [comment]
That works for me!

Frank Michielli
Sent with iOS Spark!📱

**Comment by dirkwhoffmann** [comment]
Houston, we have a problem. 😬

Background: The current prototype scans the directors for files matching *.prg and converts them to PRG files on the C64 disk (same for *.rel, etc.). This is pretty nice so far. However, there is a distinct PRG format for storing C64 programs and this format contains two additional bytes encoding the load address. If we urge the user to add a .prg suffix to all files that are supposed to be encoded as a PRG on the C64 disk, this leads to total confusion. We need another solution.

**Comment by bluecursor** [comment]
I don’t think so. I know my practice is always to add .prg and your drop down window says .prg, so I don’t see it as a big problem.

Unless you want to do somekind of dir scan and clean first before loading in the dir?


Frank Michielli
Sent with iOS Spark!📱

**Comment by dirkwhoffmann** [comment]
I've noticed that the disk encoder can only handle PRG files if the source is not a D64 or G64 file. This means that we can forget about REL, SEQ, and USR files anyway. 

I've reworked the code (again) and the new implementation works as follows: If a directory is dragged in, the emulator scans it for PRG files (files complying to the PRG format which means they encode their load address in the first two bytes). All PRG files are put in a freshly created disk. 

E.g., if the following directory (which contains real PRG files, not just files with a .prg suffix) is dragged in...

<img width="454" alt="Bildschirmfoto 2020-09-15 um 11 09 30" src="https://user-images.githubusercontent.com/12561945/93191564-f5afd700-f744-11ea-881a-340b47933f79.png">

... it looks as follows in the emulator:

<img width="857" alt="Bildschirmfoto 2020-09-15 um 11 11 10" src="https://user-images.githubusercontent.com/12561945/93191591-ff393f00-f744-11ea-93ce-84b010695872.png">
